src/sql-parser/src/lexer/buf.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::parser::ParserError;
+
+/// A cursor over a string with a variety of lexing convenience methods.
+pub struct LexBuf<'a> {
+    buf: &'a str,
+    pos: usize,
+}
+
+impl<'a> LexBuf<'a> {
+    /// Creates a new lexical buffer.
+    ///
+    /// The internal cursor is initialized to point at the start of `buf`.
+    pub fn new(buf: &'a str) -> LexBuf<'a> {
+        LexBuf { buf, pos: 0 }
+    }
+
+    /// Returns the next character in the buffer, if any, without advancing the
+    /// internal cursor.
+    pub fn peek(&self) -> Option<char> {
+        self.buf[self.pos..].chars().next()
+    }
+
+    /// Returns the next character in the buffer, if any, advancing the internal
+    /// cursor past the character.
+    ///
+    /// It is safe to call `next` after it returns `None`.
+    pub fn next(&mut self) -> Option<char> {
+        let c = self.peek();
+        if let Some(c) = c {
+            self.pos += c.len_utf8();
+        }
+        c
+    }
+
+    /// Returns a slice containing the next `n` characters in the buffer.
+    ///
+    /// Returns `None` if there are not at least `n` characters remaining.
+    /// Advances the internal cursor `n` characters, or to the end of the buffer
+    /// if there are less than `n` characters remaining.
+    pub fn next_n(&mut self, n: usize) -> Option<&'a str> {
+        let start = self.pos;
+        for _ in 0..n {
+            self.next()?;
+        }
+        Some(&self.buf[start..self.pos])
+    }
+
+    /// Returns the previous character in the buffer, positioning the internal
+    /// cursor to before the character.
+    ///
+    /// The next call to `LexBuf::next` will return the same character.
+    pub fn prev(&mut self) -> Option<char> {
+        if let Some(c) = self.buf.chars().rev().next() {
+            self.pos -= c.len_utf8();
+            Some(c)
+        } else {
+            None
```

Comment:
Should this be a panic? Is there ever a case where this can be called correctly? I mean, it seems that prev should only ever be called <= times than next is called, and so if something is calling too much, it's a bug in the lexer.

Reply:
Certainly works for me. (Panics stress some folks out. I like the eager panic approach though.)